test/api/model/web_auth_test.dart

```diff
@@ -0,0 +1,97 @@
+void main() {
+  group('WebAuthPayload', () {
+    const otp = '186f6d085a5621ebaf1ccfc05033e8acba57dae03f061705ac1e58c402c30a31';
+    final encryptedApiKey = debugEncodeApiKey(eg.selfAccount.apiKey, otp);
+    final wellFormed = Uri.parse(
+      'zulip://login?otp_encrypted_api_key=$encryptedApiKey'
+        '&email=self%40example&user_id=1&realm=https%3A%2F%2Fchat.example%2F');
+
+    test('basic happy case', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.equals(1)
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    // TODO(server-5) remove
+    test('legacy: no userId', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('user_id');
+      final payload = WebAuthPayload.parse(
+        wellFormed.replace(queryParameters: queryParams));
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.isNull()
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    test('parse fails when an expected field is missing', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('email');
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when otp_encrypted_api_key is wrong length', () {
+      final queryParams = {...wellFormed.queryParameters}
+        ..['otp_encrypted_api_key'] = 'asdf';
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when host is not "login"', () {
+      final input = wellFormed.replace(host: 'foo');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when scheme is not "zulip"', () {
+      final input = wellFormed.replace(scheme: 'https');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('decodeApiKey fails when otp is wrong length', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(() => payload.decodeApiKey('asdf')).throws<FormatException>();
+    });
+  });
+
+  group('generateOtp', () {
+    test('smoke, and check all 256 byte values are used', () {
+      const n = 216;
+      final manyOtps = List.generate(n, (_) => generateOtp());
+
+      final bytesThatAppear = <int>{};
+      for (final otp in manyOtps) {
+        final bytes = hex.decode(otp);
+        check(bytes.length).equals(32);
+        bytesThatAppear.addAll(bytes);
+      }
+
+      final byteValues = List.generate(256, (index) => index);
```

Comment:
nit: can use Iterable.generate instead of allocating a List, and then in fact can leave out the callback argument:
```suggestion
      final byteValues = Iterable.generate(256);
```

Then probably clearest to just inline that.